### DISCUSSION TITLE
Can alpineJS's @entangle can keep track of complex php arrays (Livewire class property)?

### DISCUSSION THREAD

**Discussion by thebeautyofcoding** [alpinejs/alpine#3401] [1. Help]
I am trying to build a phone keyboard, which should allow me to choose simple numbers when simply clicking, and also choosing single letters from a lettergroup, when holding the mouse down for longer, something like 500 ms. The number should be displayed in the input. With only Livewire it works, but I am unable to implement the logic for mouse click hold + normal click, or? 
This is how  the UI looks:
![image](https://user-images.githubusercontent.com/76845042/216405726-b09d6670-6538-40b2-a8c6-c1e72a20b8b1.png)

![image](https://user-images.githubusercontent.com/76845042/216405824-3b72c702-df34-49e4-b7ce-eb13b6042cd8.png)

At the moment, I am able to click on a number and see a little div with its corresponding lettergroup. But, this should only happen when holding the click for something like 500 ms. By simply clicking, I want to pick the number rather than a letter.

My Livewire class looks like this:

    <?php
    
    namespace App\Http\Livewire\Pages;
    
    use Livewire\Component;

    class DialerPage extends Component
    {

    /**
     * @property string $number
     */
    public $number = '';

    public array $phoneKeys = [
        2 => [
            'letterGroup' => ['A', 'B', 'C'],
            'shortClick' => false,
            'longClick' => false,
            'number' => 2
        ],
        3 => [
            'letterGroup' => ['D', 'E', 'F'],
            'shortClick' => false,
            'longClick' => false,
            'number' => 3
        ],
        4 => [
            'letterGroup' => ['G', 'H', 'I'],
            'shortClick' => false,
            'longClick' => false,
            'number' => 4
        ],
        5 => [
            'letterGroup' => ['J', 'K', 'L'],
            'shortClick' => false,
            'longClick' => false,
            'number' => 5
        ],
        6 => [
            'letterGroup' => ['M', 'N', 'O'],
            'shortClick' => false,
            'longClick' => false,
            'number' => 6
        ],
        7 => [
            'letterGroup' => ['P', 'Q', 'R', 'S'],
            'shortClick' => false,
            'longClick' => false,
            'number' => 7
        ],
        8 => [
            'letterGroup' => ['T', 'U', 'V'],
            'shortClick' => false,
            'longClick' => false,
            'number' => 8
        ],
        9 => [
            'letterGroup' => ['W', 'X', 'Y', 'Z'],
            'shortClick' => false,
            'longClick' => false,
            'number' => 9
        ],

    ];

    public function shortClickFn($number)
    {

        $this->phoneKeys[$number]['shortClick'] = true;
        $this->phoneKeys[$number]['longClick'] = false;
    }

    public function longClickFn($number)
    {

        $this->phoneKeys[$number]['shortClick'] = false;
        $this->phoneKeys[$number]['longClick'] = true;
    }

    /** Update number when a button on the dialer is clicked
     * @param string $number
     */
    public function updateInputNumber($number = '999999')
    {
        $this->number = $this->number . $number;
    }

    public function render()
    {
        return view(
            'livewire.pages.dialer-page',
            [

                'number' => $this->number
            ]
        );
    }
}
You can see the complex array there. What I want to archieve is the following. Whenever I click on a number for long enough, I want to alter the "longClick" property of the array and set it to true, so that I am able to display its letterGroup conditionally. I also have 2 methods shortClickFn and longClickFn that should alter these array properties, and I think they do, because I have a Laravel Debug Bar and I can see that this works. See Image:
![image](https://user-images.githubusercontent.com/76845042/216407828-b48650ab-1a45-4d31-b048-9da53c0c7273.png)
I clicked on number 2 long enough, and the corresponding property of the array got altered correctly: longClick=>true

Till now everything seems to work. Let's have a look at my template, where I am also using AlpineJS:

    `<div class="px-6">
        <div class="container flex justify-center">
            <div class="relative flex justify-between w-full">
                <input type="text" wire:model="number" class="h-14 w-full text-center text-xl font-bold rounded z-0 focus:shadow focus:outline-none" placeholder="Number..">
                <div class="absolute top-2 right-2">
                    <button class="material-icons h-10 w-10 content-end text-gray-500 rounded-lg bg-red-500 hover:bg-red-600">backspace</button>
                </div>
            </div>
        </div>
        <div class="container flex justify-center items-center w-full px-6 mt-8">
            <div class="grid grid-cols-3 gap-4">
              
                <button class="flex justify-center h-14 w-14 text-red rounded-lg bg-red-500 hover:bg-red-600 text-4xl font-bold">1</button>
                <template class="" x-data="{keys:@entangle('phoneKeys'), holding:false,timeout:null }" x-for="(details,index ) in Object.values(keys)" :key="index">
                    <div x-on:mousedown="holding = true;timeout=setTimeout(() => { $wire.longClickFn(details.number); console.log(keys,'keys') ;}, 500)" x-on:mouseup.prevent="holding = false; clearTimeout(timeout); $wire.shortClickFn(details.number);" x-on:mouseleave="holding = false;
                    console.log(details); clearTimeout(timeout)">
                        <button class=" h-14 w-14 content-end text-red rounded-lg bg-red-500 hover:bg-red-600 text-4xl font-bold flex flex-col justify-start items-center">
    
                            <div class="flex justify-start flex-col" x-text="details.number"></div>
                            <span x-text="(Object.values(details.letterGroup)).toString()" class="text-gray-500 text-sm font-bold"></span>
                        </button>
                        <template x-show="holding===true" x-for=" (letter,index) in (Object.values(details.letterGroup))" class=" absolute z-50 whitespace-nowrap left-[-50%] h-full bg-white flex justify-center items-center border-2 rounded ">
    
                            <div x-on:click="console.log( 'GROUP', letter,(Object.values(details.letterGroup)))" x-text="letter" class="transition m-1 p-1 bg-gray-200 flex justify-center items-center text-red rounded-lg bg-red-500 hover:bg-gray-500 text-2xl h-8 w-8 shadow font-bold">
                            </div>
    
                        </template>
    
                    </div>
                </template>
    
                <button class=" h-14 w-14 content-end text-red rounded-lg bg-red-500 hover:bg-red-600 text-4xl font-bold">*</button>
                <button wire:click="updateInputNumber({{ $number }})" class="h-14 w-14 content-end text-red rounded-lg bg-red-500 hover:bg-red-600 text-4xl font-bold">0</button>
                <button class=" h-14 w-14 content-end text-red rounded-lg bg-red-500 hover:bg-red-600 text-4xl font-bold">#</button>
            </div>
        </div>
    </div>`
I am using AlpineJS, so I can implement the logic for mouse click hold via setTimeout etc. Since I am using Alpine, I made use of @entangle(), and here the problem occurs. Although the php array variable $phoneKeys correctly updates, @entangle('phoneKeys') does not reflect these changes. Am I doing something wrong? Iam totally desperate :( Thanks ;)

**Comment by ekwoka** [comment]
I think the issue arrises in your x-data on the template...

I think.

I've never seen anyone do that before.

similarly you have `x-show` on a template which doesn't do anything else...
